homeassistant/components/notify/html5.py

```diff
@@ -129,7 +180,108 @@ def post(self, request):
+    name = 'api:notify.html5/callback'
+
+    def __init__(self, hass, registrations):
+        """Init HTML5PushCallbackView."""
+        super().__init__(hass)
+        self.registrations = registrations
+
+    def decode_jwt(self, token):
+        """Find the registration that signed this JWT and return it."""
+        import jwt
+
+        # 1.  Check claims w/o verifying to see if a target is in there.
+        # 2.  If target in claims, attempt to verify against the given name.
+        # 2a. If decode is successful, return the payload.
+        # 3.  If no target, loop through all registrations & attempt decode.
+        # 3a. If a registration successfully decodes, return the payload.
+        # 4.  Unable to decode the JWT, return False.
+
+        target_check = jwt.decode(token, verify=False)
+        if target_check[ATTR_TARGET] in self.registrations.keys():
+            possible_target = self.registrations[target_check[ATTR_TARGET]]
+            key = possible_target[ATTR_SUBSCRIPTION][ATTR_KEYS][ATTR_AUTH]
+            try:
+                return jwt.decode(token, key)
+            except jwt.exceptions.DecodeError:
+                pass
+
+        for reg in self.registrations.values():
+            try:
+                key = reg[ATTR_SUBSCRIPTION][ATTR_KEYS][ATTR_AUTH]
+                return jwt.decode(token, key)
+            except jwt.exceptions.DecodeError:
+                continue
+        return False
+
+    # The following is based on code from Auth0
+    # https://auth0.com/docs/quickstart/backend/python
+    # pylint: disable=too-many-return-statements
+    def check_authorization_header(self, request):
+        """Check the authorization header."""
+        import jwt
+        auth = request.headers.get('Authorization', None)
+        if not auth:
+            return self.json_message('Authorization header is expected',
+                                     status_code=HTTP_UNAUTHORIZED)
+
+        parts = auth.split()
+
+        if parts[0].lower() != 'bearer':
+            return self.json_message('Authorization header must '
+                                     'start with Bearer',
+                                     status_code=HTTP_UNAUTHORIZED)
+        elif len(parts) != 2:
+            return self.json_message('Authorization header must '
+                                     'be Bearer token',
+                                     status_code=HTTP_UNAUTHORIZED)
+
+        token = parts[1]
+        try:
+            payload = self.decode_jwt(token)
+        except jwt.exceptions.InvalidTokenError:
+            return self.json_message('token is invalid',
+                                     status_code=HTTP_UNAUTHORIZED)
+        return payload
+
+    def post(self, request):
+        """Accept the POST request for push registrations event callback."""
+        auth_check = self.check_authorization_header(request)
+        if not isinstance(auth_check, dict):
+            return auth_check
```

Comment:
I believe this logic is okay. I want to return something other than a dictionary (which is the type that comes out of `decode_jwt`/`jwt.decode`) and if it is something other than a dictionary, return the response of `auth_check`, since that will contain the `json_message`.

Let me know